promql/engine.go

```diff
@@ -1048,6 +1105,33 @@ func (ev *evaluator) eval(expr Expr) Value {
 			panic(fmt.Errorf("cannot do range evaluation of matrix selector"))
 		}
 		return ev.matrixSelector(e)
+
+	case *SubqueryExpr:
+		newEv := &evaluator{
+			endTimestamp: ev.endTimestamp - int64(e.Offset/time.Millisecond),
+			interval:     int64(getGlobalEvaluationInterval() / time.Millisecond),
+			ctx:          ev.ctx,
+			maxSamples:   ev.maxSamples,
+			logger:       ev.logger,
+		}
+		if e.StepExists {
+			newEv.interval = int64(e.Step / time.Millisecond)
+		}
+
+		// Calculate number of steps of the subquery evaluation.
+		rangeMillis := (int64(e.Range / time.Millisecond))
+		numSteps := (rangeMillis + (newEv.interval / 2)) / newEv.interval
+		if ev.startTimestamp < (newEv.endTimestamp - rangeMillis) {
+			// If the original evaluator has startTimestamp before subquery startTimestamp,
+			// have starting step on/before that.
+			numSteps = (newEv.endTimestamp - ev.startTimestamp + (newEv.interval / 2)) / newEv.interval
+		}
+
+		// Align the startTimestamp with the subquery step.
+		correctedRange := newEv.interval * numSteps
+		newEv.startTimestamp = newEv.endTimestamp - correctedRange
```

Comment:
This is aligning relative to the end time, not the step

Reply:
I am a little confused about the step now. What I understood was, the start time should be integral steps away from end time.

What do we mean by aligning with the step?

Reply:
If the step is say 5s, then the evals should be at 00:00:00, 00:00:05, 00:00:10 etc. irrelevant of what the start/end time are for the query.

Reply:
Subquery starts at time 0 always?

Reply:
Does it mean 
```
startTimestamp = 0
endTimestamp = numSteps * interval
```

Reply:
The start time will be the first aligned step at or after the start time.

Reply:
What is aligned step here? Also, shouldn't it be like the end time of subquery is the first aligned step on or before the end time, and then we take the steps back?

Reply:
The evaluator works from the start to the end, so I'd expect it to work from the start. We can then leave the endTimestamp as-is.

Reply:
Ignore my previous comment. Now things are more clear. I did not realise that we are loading the entire matrix at once.

But I don't think I still understand what do you mean by aligned step. With my current understanding, the start time should be `ev.startTimestamp - offset - range`.

Reply:
The problem with that is that it'd break the invariant that query range is just syntactic sugar over query, such as if the query range step and subquery step didn't match up.

By aligning the subquery with it's step, independent of the query/query range settings, we avoid that.

Reply:
Sorry for so much confusion, what is `it` in "subquery with it's step"?

Can you give me an example of the original step (w.r.t to the variables in the code / query) with whom we want to align the subquery's start time.

Reply:
The subquery's step, `newEv.interval`